src/wp-includes/option.php

```diff
@@ -1133,6 +1145,92 @@ function delete_option( $option ) {
+ * it will return 'auto-yes'.
+ *
+ * @since 6.5.0
+ * @access private
+ *
+ * @param string           $option   The name of the option.
+ * @param mixed            $value    The value of the option to check its autoload value.
+ * @param string|bool|null $autoload The autoload value to check.
+ *                                   Accepts 'yes'|true to enable or 'no'|false to disable, or
+ *                                   'auto-yes', 'auto-no', or 'auto' for internal purposes.
+ *                                   Any other autoload value will be forced to either `auto-yes`,
+ *                                   `auto-no`, or `auto`.
+ * @return string Returns the original $autoload value if valid, or 'auto-no' if the size exceeds
+ *                the maximum allowed size, or 'auto-yes' otherwise.
+ */
+function determine_option_autoload_value( $option, $value, $autoload ) {
+
+	/**
+	 * Filters the maximum size of option value in bytes.
+	 *
+	 * @param string|bool|null $autoload autoload value to check.
+	 * @param string           $option name of the option.
+	 * @param string           $value value of the option to check its autoload value.
+	 * @return string|bool|null $autoload autoload value to set. Accepted strings 'yes', 'no', auto-yes', 'auto-no' to enable.
+	 *
+	 * @since 6.5.0
+	 */
+	$autoload = apply_filters( 'option_autoload_value', $autoload, $option, $value );
+
+	// Check if autoload is a boolean.
+	if ( is_bool( $autoload ) ) {
+		return $autoload ? 'on' : 'off';
+	}
+
+	switch ( $autoload ) {
+		case 'yes':
+			return 'on';
+		case 'no':
+			return 'off';
+	}
+
+	/**
+	 * Allows to determine the default autoload value for an option where no explicit value is passed.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param bool|null $default The default autoload value to set. Returning true will be set as 'auto-on' in the
+	 *                           database, false will be set as 'auto-off', and null will be set as 'auto'.
+	 * @param string    $option  The passed option name.
+	 * @param mixed     $value   The passed option value to be saved.
+	 *
+	 * @retun bool|null if true returned $autoload will be set to 'auto-on' and false will set 'auto-off'.
+	 *        null will allow the size of the $autoload to be set 'auto-off' for large values.
+	 */
+
+	$autoload = apply_filters( 'wp_default_autoload_value', null, $autoload, $option, $value );
+	if ( is_bool( $autoload ) ) {
+		return $autoload ? 'auto-on' : 'auto-off';
+	}
+
+	/**
+	 * Filters the maximum size of option value in bytes.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param int    $max_option_size The option-size threshold, in bytes. Default 150000.
+	 * @param string $option          The name of the option.
+	 */
+	$max_option_size = (int) apply_filters( 'wp_max_autoloaded_option_size', 150000, $option );
+	$size            = ! empty( $value ) ? strlen( $value ) : 0;
+
+	if ( $size > $max_option_size ) {
+		return 'auto-off';
+	}
+
+	return 'auto-on';
```

Comment:
See https://github.com/WordPress/wordpress-develop/pull/5671#discussion_r1454011002, please move the size logic into a separate function hooked into the `wp_default_autoload_value` filter.
```suggestion
	return 'auto';
}

/**
 * Filters the default autoload value to disable autoloading if the option value is too large.
 *
 * @since 6.5.0
 * @access private
 *
 *@param bool|null $default The default autoload value to set. Returning true will be set as 'auto-on' in the
 *                           database, false will be set as 'auto-off', and null will be set as 'auto'.
 * @param string    $option  The passed option name.
 * @param mixed     $value   The passed option value to be saved.
 * @return bool|null Potentially modified $default.
 */
function wp_filter_default_autoload_value_via_option_size( $default, $option, $value ) {
	/**
	 * Filters the maximum size of option value in bytes.
	 *
	 * @since 6.5.0
	 *
	 * @param int    $max_option_size The option-size threshold, in bytes. Default 150000.
	 * @param string $option          The name of the option.
	 */
	$max_option_size = (int) apply_filters( 'wp_max_autoloaded_option_size', 150000, $option );
	$size            = ! empty( $value ) ? strlen( $value ) : 0;

	if ( $size > $max_option_size ) {
		return false;
	}

	return $default;
```

Then, you can add the `wp_filter_default_autoload_value_via_option_size()` function as a callback to the filter in `wp-includes/default-filters.php`.